src/marlin_pc/constraints.rs

```diff
@@ -0,0 +1,1978 @@
+    PG::G1Var: ToConstraintFieldGadget<<CycleE::E1 as PairingEngine>::Fr>,
+    PG::G2Var: ToConstraintFieldGadget<<CycleE::E1 as PairingEngine>::Fr>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    #[tracing::instrument(target = "r1cs", skip(self))]
+    fn to_constraint_field(&self) -> R1CSResult<Vec<FpVar<<CycleE::E1 as PairingEngine>::Fr>>> {
+        let mut res = Vec::new();
+
+        let mut g_gadget = self.g.to_constraint_field()?;
+        let mut h_gadget = self.h.to_constraint_field()?;
+        let mut beta_h_gadget = self.beta_h.to_constraint_field()?;
+
+        res.append(&mut g_gadget);
+        res.append(&mut h_gadget);
+        res.append(&mut beta_h_gadget);
+
+        if self.degree_bounds_and_shift_powers.as_ref().is_some() {
+            let list = self.degree_bounds_and_shift_powers.as_ref().unwrap();
+            for (_, d_gadget, shift_power) in list.iter() {
+                let mut d_elems = d_gadget.to_constraint_field()?;
+                let mut shift_power_elems = shift_power.to_constraint_field()?;
+
+                res.append(&mut d_elems);
+                res.append(&mut shift_power_elems);
+            }
+        }
+
+        Ok(res)
+    }
+}
+
+/// Var for the verification key of the Marlin-KZG10 polynomial commitment scheme.
+#[allow(clippy::type_complexity)]
+pub struct PreparedVerifierKeyVar<
+    CycleE: CycleEngine,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+> where
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    /// Generator of G1.
+    pub prepared_g: Vec<PG::G1Var>,
+    /// Generator of G2.
+    pub prepared_h: PG::G2PreparedVar,
+    /// Generator of G1, times first monomial.
+    pub prepared_beta_h: PG::G2PreparedVar,
+    /// Used for the shift powers associated with different degree bounds.
+    pub prepared_degree_bounds_and_shift_powers: Option<
+        Vec<(
+            usize,
+            FpVar<<CycleE::E1 as PairingEngine>::Fr>,
+            Vec<PG::G1Var>,
+        )>,
+    >,
+    /// Indicate whether or not it is a constant allocation (which decides whether or not shift powers are precomputed)
+    pub constant_allocation: bool,
+    /// If not a constant allocation, the original vk is attached (for computing the shift power series)
+    pub origin_vk: Option<VerifierKeyVar<CycleE, PG>>,
```

Comment:
Why is `constant_allocation` different from `is_constant()` as implemented in `R1CSVar`?